### ISSUE TITLE
Should set up BIRD for graceful restart

### ISSUE THREAD

**Issue by fasaxc** [projectcalico/calico#569]
BIRD doesn't turn on graceful restart by default, but it'd probably something that we want for BGP.  

It prevents BIRD from flapping the dataplane after it is restarted.

To turn it on, we need to set the "-R" flag when we execute BIRD.  On ubuntu, it looks like we could do that by dropping -R into the /etc/bird/envvars in the BIRD_ARGS variable.

I'm not sure how this gets configured, maybe it's just installed by upstream right now?

**Comment by alexwlchan** [comment]
The `/etc/bird/envvars` file looks like it's copied into place as part of the bird dpkg.

(Source: opening up the .deb file from http://packages.ubuntu.com/trusty/bird)

**Comment by ed-harrison** [comment]
I notice in the calico-docker BIRD config, they just set the "graceful restart" option in the kernel and BGP sections of the configuration.  Presumably we can just do that?

**Comment by DaveLangridge** [comment]
Added this config on calico-rh01 and rh02.

**Comment by matthewdupre** [comment]
@DaveLangridge Did it work?

**Comment by DaveLangridge** [comment]
I don't think it made things worse. No detailed testing of neighbors flapping.

**Comment by matthewdupre** [comment]
I had a look at this myself:

- you have to set -R in BIRD_ARGS - 'graceful restart' on its own does nothing, and
- it only applies to unplanned restarts (e.g. segfaults), not planned ones (such as service restarts).

**Comment by lxpollitt** [comment]
@robbrockbank I think you've looked into BIRD graceful restart quite a bit.  Does what @matthewdupre says above fit with your understanding too?  Does this mean that for confd driven configuration updates BIRD will not even attempt graceful restart?

**Comment by lxpollitt** [comment]
@djosborne may also be able to comment, given his prior BIRD is the word knowledge.

**Comment by robbrockbank** [comment]
In calico-docker we do run BIRD with the -R flag as well - that is required.  Restarting BIRD with a SIGHUP will invoke graceful restart, but stopping and restarting the service will not.  IIRC correctly a service restart is fine, but I'd need to check to be sure.

We also hit a couple of issues:

1) A bug in BIRD whereby we were using templates and unnamed protocols.  Graceful restart does not work with this combination.  Fix was simply to make sure the protocol derived from the template uses a unique name for each (in calico-docker we name it based on the peer IP address).

So the resulting config will be of the form:

```
template bgp bgp_template {
 ...blah
}

protocol bgp <unique name> from bgp_template {
  neighbor <addr> as <as>
}
```

(Originally we didn't include the <unique name> and that prevented GR from working).

2) We originally included a timestamp in the generated BIRD config.  The upshot of this is that any configuration change to etcd resulted in confd generating new BIRD configuration for every node and restarting multiple BIRD instances at the same time.  In this scenario GR will not work.  Solution here was to remove the timestamp - confd will not restart BIRD if the config has not actually changed.

----------

On the latter point, there is still an issue using confd for the BIRD templates.  Namely, if there is a genuine update that results in multiple BIRD configs changing then there will still be multiple simultaneous BIRD restarts and GR is not guaranteed to work.  In calico-docker this will occur when another node is added to the full node-to-node mesh, and also when another IP pool is added.  I'm not sure this would be easy to resolve without scrapping the use of confd - unless there is a way to stagger changes.

**Comment by matthewdupre** [comment]
I tried a service restart and it wasn't graceful.

I think this is the main problem - I'd have thought our biggest use case for BIRD graceful restart is software updates, and while I don't know if those exactly use service restart or stop/start, I suspect they do.

The templates are interesting.

**Comment by eepyaich** [comment]
From the perspective of the Calico solution having graceful restart, I think we need to address this now, i.e. work out what we need to change in the BIRD config, test it and document it.

Assigning to @matthewdupre as he looked at this most recently.

**Comment by matthewdupre** [comment]
Fixed in #764.
